include/oneapi/dpl/pstl/execution_impl.h

```diff
@@ -105,25 +78,135 @@ template <typename _ExecutionPolicy>
+constexpr auto
+__is_parallelization_preferred()
+{
+    return ::std::conjunction<decltype(::std::decay_t<_ExecutionPolicy>::__allow_parallel()),
+                              __internal::__is_random_access_iterator_t<_IteratorTypes...>>();
+}
+
+//------------------------------------------------------------------------
+// backend selector with tags
+//------------------------------------------------------------------------
+
+#if _ONEDPL_PAR_BACKEND_TBB
+using __par_backend_tag = __tbb_backend_tag;
+#elif _ONEDPL_PAR_BACKEND_OPENMP
+using __par_backend_tag = __omp_backend_tag;
+#elif _ONEDPL_PAR_BACKEND_SERIAL
+using __par_backend_tag = __serial_backend_tag;
+#else
+#    error "Parallel backend was not specified"
+#endif
+
+template <class _IsVector>
+struct __serial_tag
+{
+    using __is_vector = _IsVector;
+};
+
+template <class _IsVector>
+struct __parallel_tag
+{
+    using __is_vector = _IsVector;
+    // backend tag can be change depending on
+    // TBB availability in the environment
+    using __backend_tag = __par_backend_tag;
+};
+
+struct __parallel_forward_tag
+{
+    using __is_vector = ::std::false_type;
+    // backend tag can be change depending on
+    // TBB availability in the environment
+    using __backend_tag = __par_backend_tag;
+};
+
+template <class _IsVector, class... _IteratorTypes>
+using __tag_type = ::std::conditional_t<
+    __internal::__is_random_access_iterator_v<_IteratorTypes...>, __parallel_tag<_IsVector>,
+    ::std::conditional_t<__is_forward_iterator_v<_IteratorTypes...>, __parallel_forward_tag, __serial_tag<_IsVector>>>;
+
+template <class _ExecutionPolicy, class... _IteratorTypes>
+constexpr ::std::enable_if_t<
+    ::std::is_same_v<::std::decay_t<_ExecutionPolicy>, oneapi::dpl::execution::sequenced_policy>,
+    __serial_tag<std::false_type>>
+__select_backend()
+{
+    return {};
+}
+
+template <class _ExecutionPolicy, class... _IteratorTypes>
+constexpr ::std::enable_if_t<
+    ::std::is_same_v<::std::decay_t<_ExecutionPolicy>, oneapi::dpl::execution::unsequenced_policy>,
+    __serial_tag<__internal::__is_random_access_iterator<_IteratorTypes...>>>
+__select_backend()
+{
+    return {};
+}
+
+template <class _ExecutionPolicy, class... _IteratorTypes>
+constexpr ::std::enable_if_t<
+    ::std::is_same_v<::std::decay_t<_ExecutionPolicy>, oneapi::dpl::execution::parallel_policy>,
+    __tag_type<std::false_type, _IteratorTypes...>>
+__select_backend()
+{
+    return {};
+}
+
+template <class _ExecutionPolicy, class... _IteratorTypes>
+constexpr ::std::enable_if_t<
+    ::std::is_same_v<::std::decay_t<_ExecutionPolicy>, oneapi::dpl::execution::parallel_unsequenced_policy>,
+    __tag_type<__internal::__is_random_access_iterator<_IteratorTypes...>, _IteratorTypes...>>
+__select_backend()
 {
-    return __internal::__lazy_and(__exec.__allow_parallel(),
-                                  __internal::__is_random_access_iterator_t<_IteratorTypes...>());
+    return {};
 }
 
+template <class _Tag>
```

Comment:
Am I right the following traits (lines 174-208) are used just for static asserts?

Reply:
Yes, it's correct.

Reply:
But also we have
```C++
template <class _Tag>
struct __is_backend_tag;

template <typename _BackendTag>
struct __is_backend_tag<__hetero_tag<_BackendTag>> : ::std::true_type
{
};
```

in the separate file.

Reply:
> Yes, it's correct.

In that case I would say we dont need so many definitions which just confused everyone who is trying to read the code.

We told about assert just for "general" tag for algo pattern. That tag, essentially, is a serial tag (for the standard host policies). We are dispatching to that pattern in case of `seq `and `unseq` policies, and also in case of "unknown" policy (if someone wants to add an overload for the pattern to customize implementation). So, to control "dispatching" we need just one trait like is_seq_or_unseq_or_unknown_policy or/and non-random access iterators...
Other words, it is enough to check that a tag is a __serial_tag and checking expression should call like "is_serial_tag_v".... I will discuss it with Ruslan one more time.

Reply:
@MikeDvorskiy  And what about  our `__parallel_tag` ? Can we assume that it is serial tag too in our `static_assert` checks?

Reply:
@SergeyKopienko, I don't catch your last question...